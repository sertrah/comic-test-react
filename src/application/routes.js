import Home from "presentation-layer/home";

import { ROUTER_PATH_LIST } from "./constants";

const routes = [
  {
    path: ROUTER_PATH_LIST.default,
    component: Home,
  }
];

export default routes;
