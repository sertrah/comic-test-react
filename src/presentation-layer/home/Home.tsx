import React, { FunctionComponent, useMemo } from "react";
import { useQuery } from "react-query";
import { useNotification } from "application/hooks";

import { comicService } from "application/services/Comic.service";
import { StarRating, Card, Loader } from "presentation-layer/components";
import { commonHelper } from "application/helper";
import { Comic } from "application/models/Comic.model";

// I did this part because it did not work for me to consume the resource either with http or https or proxys
const cardFake = {
  month: 8,
  num: 628,
  link: "",
  year: 2009,
  news: "",
  safe_title: "Psychic",
  transcript:
    "Man: I'm psychic, you know. Woman: There's no such thing. Man: Okay, think of a number from one to one hundred. Woman: Okay. Man: 43. Woman: Holy shit! Man: I try not to let it affect my life too much. Woman: Wait, I can't believe this. Man: Don't worry about it. Forget I said anything. Woman: But-- Man: Let's get to the movie. Woman: I, uh... Ok, sure. Narrator: This trick may only work 1% of the time, but when it does, it's totally worth it. {{Title text: You can do a lot better than 1% if you start keeping track of the patterns in what numbers people pick.}}",
  alt:
    "You can do a lot better than 1% if you start keeping track of the patterns in what numbers people pick.",
  img: "https://imgs.xkcd.com/comics/psychic.png",
  title: "Psychic",
  day: 26,
};

const Home: FunctionComponent = () => {
  const [starsSelected, setStarsSelected] = React.useState<number>(2);
  const randomNumber = useMemo(() => commonHelper.getRandomNumber(700), []);
  const notify = useNotification();

  const { data: comic, isLoading } = useQuery(
    ["comic", randomNumber],
    ({ queryKey: [, ramdonNumber] }) => comicService.getComic(ramdonNumber),
    {
      retry: 1,
      retryDelay: 3000,
      onSuccess: (comicResponse) => {
        comicService.getComicRating(comicResponse.num).then((rating) => {
          setStarsSelected(rating);
        });
      },
      onError: ({ message }) => {
        notify.error({
          title: "Comic",
          message: `${message}. You are looking at an old comic`,
        });
      },
    }
  );

  const comicValidator: Comic = useMemo(() => comic ?? cardFake, [comic]);

  const onSaveRating = (rating: number) => {
    // It is not wrong to make sure for possible errors: D
    try {
      if (comicValidator) {
        setStarsSelected(rating);
        comicService.saveComicRating(comicValidator.num, rating).then(() => {
          notify.success({
            title: "Comic",
            message: "Thanks for voting",
          });
        });
      }
    } catch (error) {
      notify.error({
        title: "Comic",
        message: `What a shame, contact support`,
      });
    }
  };

  return (
    <>
      {isLoading && <Loader />}
      {comicValidator ? (
        <Card comic={comicValidator}>
          <StarRating
            handleInputValue={onSaveRating}
            totalStars={5}
            value={starsSelected}
          />
        </Card>
      ) : (
        <h1 className="title__error">
          Ups something is wrong, please try again later
        </h1>
      )}
    </>
  );
};

export default Home;
