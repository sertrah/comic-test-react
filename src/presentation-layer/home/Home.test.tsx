import React from "react";
import { render } from "@testing-library/react";
import { QueryClientProvider, QueryClient } from "react-query";
import { NotificationProvider } from "application/contexts";
import Home from "./Home";

const queryClient = new QueryClient();

describe("Home tests", () => {


  it("renders well", () => {
    const component = render(
      <QueryClientProvider client={queryClient}>
        <NotificationProvider>
          <Home />{" "}
        </NotificationProvider>
      </QueryClientProvider>
    );
    expect(component.container).toMatchSnapshot();
  });
});
