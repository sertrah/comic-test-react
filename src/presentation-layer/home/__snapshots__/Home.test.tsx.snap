// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home tests renders well 1`] = `
<div>
  <div
    class="loader"
  >
    <div
      class="loader--container"
    >
      <div
        class="loader--item"
      />
      <p />
    </div>
  </div>
  <article
    aria-label="card containing information about comic"
    class="comic-card"
  >
    <div
      class="comic-card__img"
    >
      <img
        alt="You can do a lot better than 1% if you start keeping track of the patterns in what numbers people pick."
        src="https://imgs.xkcd.com/comics/psychic.png"
      />
    </div>
    <div
      class="comic-card__content"
    >
      <span
        class="comic-card__date"
      >
        26/ 8/ 2009
      </span>
      <h1
        class="comic-card__title"
      >
        Psychic
      </h1>
      <p
        class="comic-card__text"
      >
        Man: I'm psychic, you know. Woman: There's no such thing. Man: Okay, think of a number from one to one hundred. Woman: Okay. Man: 43. Woman: Holy shit! Man: I try not to let it affect my life too much. Woman: Wait, I can't believe this. Man: Don't worry about it. Forget I said anything. Woman: But-- Man: Let's get to the movie. Woman: I, uh... Ok, sure. Narrator: This trick may only work 1% of the time, but when it does, it's totally worth it. {{Title text: You can do a lot better than 1% if you start keeping track of the patterns in what numbers people pick.}}
      </p>
      <div
        class="comic-card__action"
      >
        <div
          class="rating"
        >
          <label>
            <input
              name="stars"
              type="radio"
              value="1"
            />
            <span
              class="icon"
            >
              ★
            </span>
          </label>
          <label>
            <input
              checked=""
              name="stars"
              type="radio"
              value="2"
            />
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
          </label>
          <label>
            <input
              name="stars"
              type="radio"
              value="3"
            />
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
          </label>
          <label>
            <input
              name="stars"
              type="radio"
              value="4"
            />
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
          </label>
          <label>
            <input
              name="stars"
              type="radio"
              value="5"
            />
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
            <span
              class="icon"
            >
              ★
            </span>
          </label>
        </div>
      </div>
    </div>
  </article>
   
</div>
`;
